src/evo/netinfo.cpp

```diff
@@ -24,6 +25,131 @@ const CChainParams& MainParams()
+        [](auto&& lhs, auto&& rhs) -> bool {
+            if constexpr (std::is_same_v<decltype(lhs), decltype(rhs)>) {
+                return lhs == rhs;
+            }
+            return false;
+        },
+        m_data, rhs.m_data);
+}
+
+bool NetInfoEntry::operator<(const NetInfoEntry& rhs) const
+{
+    if (m_type != rhs.m_type) return m_type < rhs.m_type;
+    return std::visit(
+        [](auto&& lhs, auto&& rhs) -> bool {
+            using T1 = std::decay_t<decltype(lhs)>;
+            using T2 = std::decay_t<decltype(rhs)>;
+            if constexpr (std::is_same_v<T1, T2>) {
+                // Both the same type, compare as usual
+                return lhs < rhs;
+            } else if constexpr (std::is_same_v<T1, std::monostate> && !std::is_same_v<T2, std::monostate>) {
+                // lhs is monostate and rhs is not, rhs is greater
+                return true;
+            } else if constexpr (!std::is_same_v<T1, std::monostate> && std::is_same_v<T2, std::monostate>) {
+                // rhs is monostate but lhs is not, lhs is greater
+                return false;
+            }
+            return false;
+        },
+        m_data, rhs.m_data);
+}
+
+std::optional<std::reference_wrapper<const CService>> NetInfoEntry::GetAddrPort() const
+{
+    if (const auto* data_ptr{std::get_if<CService>(&m_data)}; m_type == NetInfoType::Service && data_ptr) {
+        ASSERT_IF_DEBUG(data_ptr->IsValid());
+        return *data_ptr;
+    }
+    return std::nullopt;
+}
+
+uint16_t NetInfoEntry::GetPort() const
+{
+    return std::visit(
+        [](auto&& input) -> uint16_t {
+            using T1 = std::decay_t<decltype(input)>;
+            if constexpr (std::is_same_v<T1, CService>) {
+                return input.GetPort();
+            } else {
+                return 0;
+            }
+        },
+        m_data);
+}
+
+// NetInfoEntry is a dumb object that doesn't enforce validation rules, that is the responsibility of
+// types that utilize NetInfoEntry (MnNetInfo and others). IsTriviallyValid() is there to check if a
+// NetInfoEntry object is properly constructed.
+bool NetInfoEntry::IsTriviallyValid() const
+{
+    if (m_type == NetInfoType::Invalid) return false;
+    return std::visit(
+        [this](auto&& input) -> bool {
+            using T1 = std::decay_t<decltype(input)>;
+            if constexpr (std::is_same_v<T1, std::monostate>) {
+                // Empty underlying data isn't a valid entry
+                return false;
+            } else if constexpr (std::is_same_v<T1, CService>) {
+                // Type code should be truthful as it decides what underlying type is used when (de)serializing
+                if (m_type != NetInfoType::Service) return false;
+                // Underlying data must meet surface-level validity checks for its type
+                if (!input.IsValid()) return false;
+            } else {
+                return false;
+            }
+            return true;
```

Comment:
may be simplify it a bit?
```
            if constexpr (std::is_same_v<T1, std::monostate>) {
                // Empty underlying data isn't a valid entry
                return false;
            }
            if constexpr (std::is_same_v<T1, CService>) {
                // Type code should be truthful as it decides what underlying type is used when (de)serializing
                if (m_type != NetInfoType::Service) return false;
                // Underlying data must meet surface-level validity checks for its type
                return input.IsValid();
            }
            return false;
```

Reply:
or even better add

```
    static_assert(std::is_same_v<T1, std::monostate> || std::is_same_v<T1, CService>,
                  "type is not implemented");
```

In this case compiler tell directly if type is not supported and no need "else" option at all

Reply:
Added `static_assert`, removed extraneous `else`. Resolved in latest push!